CrocoDash/data_access/large_data_workflow/code/get_data_piecewise.py

```diff
@@ -0,0 +1,166 @@
+from pathlib import Path
+from CrocoDash import utils
+from CrocoDash.data_access.driver import get_rectangular_segment_info
+from CrocoDash.data_access.utils import load_config
+from CrocoDash.data_access import driver as dv
+import xarray as xr
+import pandas as pd
+from datetime import datetime, timedelta
+
+logger = utils.setup_logger(__name__)
+
+
+def get_data_piecewise(
+    product_name: str,
+    function_name: str,
+    date_format: str,
+    start_date: str,
+    end_date: str,
+    hgrid_path: str | Path,
+    step_days: int,
+    output_dir: str | Path,
+    boundary_number_conversion: dict,
+    preview: bool = False,
+):
+    """
+    Retrieves and saves data in piecewise chunks for each boundary over a date range.
+
+    Parameters
+    ----------
+    product_name : str
+        The name of the data product to retrieve.
+    function_name : str
+        The function to call for retrieving data.
+    date_format : str
+        The date format string (e.g., "%Y-%m-%d").
+    start_date : str
+        The start date in the specified format.
+    end_date : str
+        The end date in the specified format.
+    hgrid_path : str or Path
+        Path to the hgrid file containing the regional grid.
+    step_days : int
+        The number of days in each data chunk.
+    output_dir : str or Path
+        The directory to save the output NetCDF files.
+    boundary_number_conversion : dict
+        Dictionary mapping boundaries to their numerical identifiers.
+    preview : bool
+        Whether or not to preview the run, default is false
+
+    Raises
+    ------
+    ValueError
+        If the product or function is not found in the registry.
+
+    Returns
+    -------
+    None
+        Saves the retrieved data to the specified output directory.
+    """
+
+    ## Initialize PFD
+    ProductFunctionRegistry = dv.ProductFunctionRegistry()
+    ProductFunctionRegistry.load_functions()
+    ProductFunctionRegistry.validate_function(product_name, function_name)
+    func = ProductFunctionRegistry.functions[product_name][function_name]
+
+    # Get lat,lon information for each boundary
+    hgrid = xr.open_dataset(hgrid_path)
+    boundary_info = get_rectangular_segment_info(hgrid)
+
+    # Set up date range
+    dates = (
+        pd.date_range(start=start_date, end=end_date, freq=f"{step_days}D")
+        .to_pydatetime()
+        .tolist()
+    )
+
+    # Add the end date manually if not included
+    if dates[-1] != datetime.strptime(start_date, date_format):
+        dates.append(datetime.strptime(end_date, date_format))
+
+    num_files = len(dates) - 1
```

Comment:
The reason why num_files one less than `len(dates)` is unclear to me, so it should probably be briefly explained in a preceding comment line.